### ISSUE TITLE
Some questions for the emotion vector

### ISSUE THREAD

**Issue by ChairManMeow-SY** [leventt/surat#7]
Thank you for sharing the code! I 'm grenn in this filed and I have some question about the emotion vector. 

In the model forward function:

```
       else:
            out = torch.cat(
                (
                    out,
                    torch.cat((
                        self.mood[moodIndex.chunk(chunks=1, dim=0)],
                        self.mood[(moodIndex + 1).chunk(chunks=1, dim=0)]
                    ), dim=0).view(
                        out.size()[0], self.moodLen, 1, 1
                    ).expand(out.size()[0], self.moodLen, 64, 1)
                ),
                dim=1
            ).view(-1, 256 + self.moodLen, 64, 1)
```

here the self.mood is in size [moodSize, moodLen]. The moodSize is the count of dataset, and the moodLen is 16 as set by the paper. 

Here the first cat function will generate a tensor in shape [2, 16], is that right? Then the view confuses me because the out.size()[0] should be the number of minibatch. Should it throw an error here?

Please correct me if I misunderstand anything... And again many tahnks for this implementation!

**Comment by leventt** [comment]
This is confusing partly because it isn't clearly noted there and it gets hard for me to follow as well after a few years, even though I wrote it :)

In the paper and this implementation, data is passed as pairs of consecutive frames... This is helpful to calculate the mood loss by just shifting the mood and taking the difference. However, moodIndex is not a pair of indices but an index for the first frame in the pair. So when I index with moodIndex and (moodIndex + 1), it is exactly matching the minibatch size `out.size()[0]`. That view operation, adds a couple redundant dimensions at the end so it can be expanded/broadcasted into (N, 16, 64, 1) per mood pair in the batch.

I really should clean this code up and make it more readable at some point.

**Comment by ChairManMeow-SY** [comment]
Thank you leventt. It is so kind of you to explain it. I'll fork it and try to write some comments -- I believe that it is an excellent textbook for anyone who wants to learn audio2face.

**Comment by leventt** [comment]
much appreciated :)
